File: backend/Backend_API/Backend_API/Models/Entities/Listing.cs
using System;
using System.Collections.Generic;

namespace Backend_API.Models.Entities;

public partial class Listing
{
    public long ListingId { get; set; }

    public long LandlordId { get; set; }

    public int TypeId { get; set; }

    public int StatusId { get; set; }

    public string Title { get; set; } = null!;

    public string? Description { get; set; }

    public decimal Price { get; set; }

    public decimal Area { get; set; }

    public int? Floor { get; set; }

    public int? TotalFloors { get; set; }

    public int? MaxOccupants { get; set; }

    public int? ProvinceId { get; set; }

    public int? DistrictId { get; set; }

    public int? WardId { get; set; }

    public string StreetAddress { get; set; } = null!;

    public double? Latitude { get; set; }

    public double? Longitude { get; set; }

    public bool? IsVerified { get; set; }

    public bool? IsFeatured { get; set; }

    public bool? IsNew { get; set; }

    public bool? AllowPet { get; set; }

    public string? Image0 { get; set; }

    public string? Image1 { get; set; }

    public string? Image2 { get; set; }

    public string? Image3 { get; set; }

    public string? Image4 { get; set; }

    public string? Image5 { get; set; }

    public string? CloudinaryFolder { get; set; }

    public decimal? ElectricPrice { get; set; }

    public decimal? WaterPrice { get; set; }

    public decimal? InternetPrice { get; set; }

    public decimal? ParkingPrice { get; set; }

    public int? ViewCount { get; set; }

    public int? SaveCount { get; set; }

    public DateOnly? AvailableFrom { get; set; }

    public DateTime? ExpiredAt { get; set; }

    public DateTime? CreatedAt { get; set; }

    public DateTime? UpdatedAt { get; set; }

    public virtual ICollection<Banner> Banners { get; set; } = new List<Banner>();

    public virtual ICollection<Conversation> Conversations { get; set; } = new List<Conversation>();

    public virtual District? District { get; set; }

    public virtual ICollection<Favorite> Favorites { get; set; } = new List<Favorite>();

    public virtual ICollection<Invoice> Invoices { get; set; } = new List<Invoice>();

    public virtual User Landlord { get; set; } = null!;

    public virtual ICollection<ListingAmenity> ListingAmenities { get; set; } = new List<ListingAmenity>();

    public virtual ICollection<ListingImage> ListingImages { get; set; } = new List<ListingImage>();

    public virtual ICollection<ListingPostPackage> ListingPostPackages { get; set; } = new List<ListingPostPackage>();

    public virtual ICollection<ListingPriceHistory> ListingPriceHistories { get; set; } = new List<ListingPriceHistory>();

    public virtual ICollection<ListingVideo> ListingVideos { get; set; } = new List<ListingVideo>();

    public virtual Province? Province { get; set; }

    public virtual ICollection<Report> Reports { get; set; } = new List<Report>();

    public virtual ICollection<Review> Reviews { get; set; } = new List<Review>();

    public virtual ListingStatus Status { get; set; } = null!;

    public virtual RoomType Type { get; set; } = null!;

    public virtual ICollection<ViewHistory> ViewHistories { get; set; } = new List<ViewHistory>();

    public virtual Ward? Ward { get; set; }
}

File: lib/models/listing.dart
// lib/models/listing.dart

class Listing {
  final int listingId;
  final String title;
  final String? description;
  final double price;
  final double area;
  final int typeId;
  final String typeName;

  // địa chỉ
  final String? provinceName;
  final String? districtName;
  final String? wardName;
  final String streetAddress;
  final double? latitude;
  final double? longitude;

  // ảnh: image0 là ảnh bìa, image1-5 là gallery
  final String? image0;
  final String? image1;
  final String? image2;
  final String? image3;
  final String? image4;
  final String? image5;

  // thông tin chủ trọ
  final int? landlordId;
  final String? landlordName;
  final String? landlordAvatar;
  final String? landlordPhone;

  // giá điện nước internet xe - null = chưa có thông tin
  final double? electricPrice;
  final double? waterPrice;
  final double? internetPrice;
  final double? parkingPrice;

  // thông tin phòng
  final int? floor;
  final int? totalFloors;
  final int? maxOccupants;
  final bool allowPet;
  final DateTime? availableFrom;

  final bool isVerified;
  final bool isFeatured;
  final String statusName;
  final int? viewCount;
  final int? saveCount;

  // rating trung bình và số lượng đánh giá
  final double averageRating;
  final int reviewCount;

  final List<String> amenityNames;
  final DateTime? createdAt;
  final Map<String, dynamic>? packageInfo;

  const Listing({
    required this.listingId,
    required this.title,
    this.description,
    required this.price,
    required this.area,
    required this.typeId,
    required this.typeName,
    this.provinceName,
    this.districtName,
    this.wardName,
    required this.streetAddress,
    this.latitude,
    this.longitude,
    this.image0,
    this.image1,
    this.image2,
    this.image3,
    this.image4,
    this.image5,
    this.landlordId,
    this.landlordName,
    this.landlordAvatar,
    this.landlordPhone,
    this.electricPrice,
    this.waterPrice,
    this.internetPrice,
    this.parkingPrice,
    this.floor,
    this.totalFloors,
    this.maxOccupants,
    required this.allowPet,
    this.availableFrom,
    required this.isVerified,
    required this.isFeatured,
    required this.statusName,
    this.viewCount,
    this.saveCount,
    this.averageRating = 0,
    this.reviewCount = 0,
    required this.amenityNames,
    this.createdAt,
    this.packageInfo,
  });

  factory Listing.fromJson(Map<String, dynamic> json) {
    // hỗ trợ cả camelCase lẫn PascalCase vì backend .NET trả về PascalCase
    T? read<T>(String camel, String pascal) {
      final value = json[camel] ?? json[pascal];
      return value is T ? value : null;
    }

    double number(String camel, String pascal) {
      final value = json[camel] ?? json[pascal];
      if (value is num) return value.toDouble();
      return double.tryParse(value?.toString() ?? '') ?? 0;
    }

    double? nullableNumber(String camel, String pascal) {
      final value = json[camel] ?? json[pascal];
      if (value == null) return null;
      if (value is num) return value.toDouble();
      return double.tryParse(value.toString());
    }

    int integer(String camel, String pascal) {
      final value = json[camel] ?? json[pascal];
      if (value is num) return value.toInt();
      return int.tryParse(value?.toString() ?? '') ?? 0;
    }

    int? nullableInt(String camel, String pascal) {
      final value = json[camel] ?? json[pascal];
      if (value == null) return null;
      if (value is num) return value.toInt();
      return int.tryParse(value.toString());
    }

    bool boolean(String camel, String pascal) {
      final value = json[camel] ?? json[pascal];
      if (value is bool) return value;
      return value?.toString().toLowerCase() == 'true';
    }

    final amenitiesRaw = json['amenityNames'] ?? json['AmenityNames'];
    final packageRaw = json['packageInfo'] ?? json['PackageInfo'];
    final createdRaw = json['createdAt'] ?? json['CreatedAt'];
    final availableRaw = json['availableFrom'] ?? json['AvailableFrom'];

    return Listing(
      listingId: integer('listingId', 'ListingId'),
      title: read<String>('title', 'Title') ?? '',
      description: read<String>('description', 'Description'),
      price: number('price', 'Price'),
      area: number('area', 'Area'),
      typeId: integer('typeId', 'TypeId'),
      typeName: read<String>('typeName', 'TypeName') ?? '',
      provinceName: read<String>('provinceName', 'ProvinceName'),
      districtName: read<String>('districtName', 'DistrictName'),
      wardName: read<String>('wardName', 'WardName'),
      streetAddress: read<String>('streetAddress', 'StreetAddress') ?? '',
      latitude: nullableNumber('latitude', 'Latitude'),
      longitude: nullableNumber('longitude', 'Longitude'),
      image0: read<String>('image0', 'Image0'),
      image1: read<String>('image1', 'Image1'),
      image2: read<String>('image2', 'Image2'),
      image3: read<String>('image3', 'Image3'),
      image4: read<String>('image4', 'Image4'),
      image5: read<String>('image5', 'Image5'),
      landlordId: nullableInt('landlordId', 'LandlordId'),
      landlordName: read<String>('landlordName', 'LandlordName'),
      landlordAvatar: read<String>('landlordAvatar', 'LandlordAvatar'),
      landlordPhone: read<String>('landlordPhone', 'LandlordPhone'),
      electricPrice: nullableNumber('electricPrice', 'ElectricPrice'),
      waterPrice: nullableNumber('waterPrice', 'WaterPrice'),
      internetPrice: nullableNumber('internetPrice', 'InternetPrice'),
      parkingPrice: nullableNumber('parkingPrice', 'ParkingPrice'),
      floor: nullableInt('floor', 'Floor'),
      totalFloors: nullableInt('totalFloors', 'TotalFloors'),
      maxOccupants: nullableInt('maxOccupants', 'MaxOccupants'),
      allowPet: boolean('allowPet', 'AllowPet'),
      availableFrom: availableRaw == null
          ? null
          : DateTime.tryParse(availableRaw.toString()),
      isVerified: boolean('isVerified', 'IsVerified'),
      isFeatured: boolean('isFeatured', 'IsFeatured'),
      statusName: read<String>('statusName', 'StatusName') ?? '',
      viewCount: nullableInt('viewCount', 'ViewCount'),
      saveCount: nullableInt('saveCount', 'SaveCount'),
      averageRating: number('averageRating', 'AverageRating'),
      reviewCount: integer('reviewCount', 'ReviewCount'),
      amenityNames: amenitiesRaw is List
          ? amenitiesRaw.map((e) => e.toString()).toList()
          : const [],
      createdAt:
          createdRaw == null ? null : DateTime.tryParse(createdRaw.toString()),
      packageInfo: packageRaw is Map<String, dynamic> ? packageRaw : null,
    );
  }

  // ghép địa chỉ đầy đủ để hiển thị
  String get displayAddress {
    final parts = [
      if (streetAddress.trim().isNotEmpty) streetAddress.trim(),
      if (wardName?.trim().isNotEmpty == true) wardName!.trim(),
      if (districtName?.trim().isNotEmpty == true) districtName!.trim(),
      if (provinceName?.trim().isNotEmpty == true) provinceName!.trim(),
    ];
    return parts.isEmpty ? 'Chưa cập nhật địa chỉ' : parts.join(', ');
  }

  // lấy tất cả ảnh không null để làm gallery
  List<String> get allImages {
    return [image0, image1, image2, image3, image4, image5]
        .whereType<String>()
        .where((url) => url.trim().isNotEmpty)
        .toList();
  }

  bool get hasLocation => latitude != null && longitude != null;

  Listing copyWith({
    String? title,
    String? description,
    String? typeName,
    String? provinceName,
    String? districtName,
    String? wardName,
    String? streetAddress,
    List<String>? amenityNames,
  }) {
    return Listing(
      listingId: listingId,
      title: title ?? this.title,
      description: description ?? this.description,
      price: price,
      area: area,
      typeId: typeId,
      typeName: typeName ?? this.typeName,
      provinceName: provinceName ?? this.provinceName,
      districtName: districtName ?? this.districtName,
      wardName: wardName ?? this.wardName,
      streetAddress: streetAddress ?? this.streetAddress,
      latitude: latitude,
      longitude: longitude,
      image0: image0,
      image1: image1,
      image2: image2,
      image3: image3,
      image4: image4,
      image5: image5,
      landlordId: landlordId,
      landlordName: landlordName,
      landlordAvatar: landlordAvatar,
      landlordPhone: landlordPhone,
      electricPrice: electricPrice,
      waterPrice: waterPrice,
      internetPrice: internetPrice,
      parkingPrice: parkingPrice,
      floor: floor,
      totalFloors: totalFloors,
      maxOccupants: maxOccupants,
      allowPet: allowPet,
      availableFrom: availableFrom,
      isVerified: isVerified,
      isFeatured: isFeatured,
      statusName: statusName,
      viewCount: viewCount,
      saveCount: saveCount,
      averageRating: averageRating,
      reviewCount: reviewCount,
      amenityNames: amenityNames ?? this.amenityNames,
      createdAt: createdAt,
      packageInfo: packageInfo,
    );
  }
}
